File: psra/web/src/main/java/by/arhor/psra/config/UserRoleTokenEnhancer.java
package by.arhor.psra.config;

import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.oauth2.common.DefaultOAuth2AccessToken;
import org.springframework.security.oauth2.common.OAuth2AccessToken;
import org.springframework.security.oauth2.provider.OAuth2Authentication;
import org.springframework.security.oauth2.provider.token.TokenEnhancer;

import static java.util.stream.Collectors.toSet;

public class UserRoleTokenEnhancer implements TokenEnhancer {
  @Override
  public OAuth2AccessToken enhance(OAuth2AccessToken token, OAuth2Authentication auth) {
    final var scopes = auth
      .getUserAuthentication()
      .getAuthorities()
      .stream()
      .map(GrantedAuthority::getAuthority)
      .collect(toSet());

    ((DefaultOAuth2AccessToken) token).setScope(scopes);
    return token;
  }
}

File: psra/web/src/main/scala/by/arhor/psra/config/UserRoleTokenEnhancer.scala
package by.arhor.psra.config

import java.util.stream.Collectors.toSet

import org.springframework.security.oauth2.common.{DefaultOAuth2AccessToken, OAuth2AccessToken}
import org.springframework.security.oauth2.provider.OAuth2Authentication
import org.springframework.security.oauth2.provider.token.TokenEnhancer

class UserRoleTokenEnhancer extends TokenEnhancer {

    override
    def enhance(token: OAuth2AccessToken, auth: OAuth2Authentication): OAuth2AccessToken = {
        val scopes = auth.getUserAuthentication
          .getAuthorities
          .stream
          .map[String] { _.getAuthority }
          .collect(toSet())

        token.asInstanceOf[DefaultOAuth2AccessToken].setScope(scopes)
        token
    }

}
